control-plane/grpc/src/operations/volume/traits.rs

```diff
@@ -0,0 +1,908 @@
+            status: volume_spec_status,
+            target: match grpc_volume_spec.target {
+                Some(target) => match VolumeTarget::try_from(target) {
+                    Ok(target) => Some(target),
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => None,
+            },
+            policy: match grpc_volume_spec.policy {
+                Some(policy) => policy.into(),
+                None => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+            },
+            topology: match grpc_volume_spec.topology {
+                Some(topology) => match Topology::try_from(topology) {
+                    Ok(topology) => Some(topology),
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => None,
+            },
+            sequencer: Default::default(),
+            last_nexus_id: match grpc_volume_spec.last_nexus_id {
+                Some(id) => match NexusId::try_from(id) {
+                    Ok(volume_id) => Some(volume_id),
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => None,
+            },
+            operation: None,
+        };
+        let volume_state = VolumeState {
+            uuid: match grpc_volume_state.uuid {
+                Some(id) => match VolumeId::try_from(id) {
+                    Ok(volume_id) => volume_id,
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+            },
+            size: grpc_volume_state.size,
+            status: match nexus::NexusStatus::from_i32(grpc_volume_state.status) {
+                Some(status) => status.into(),
+                None => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+            },
+            target: match grpc_volume_state.target {
+                Some(target) => match Nexus::try_from(target) {
+                    Ok(target) => Some(target),
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => None,
+            },
+            replica_topology: match to_replica_topology_map(grpc_volume_state.replica_topology) {
+                Ok(replica_topology_map) => replica_topology_map,
+                Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+            },
+        };
+        Ok(Volume::new(volume_spec, volume_state))
+    }
+}
+
+impl TryFrom<volume::Volumes> for Volumes {
+    type Error = ReplyError;
+    fn try_from(grpc_volume_type: volume::Volumes) -> Result<Self, Self::Error> {
+        let mut volumes: Vec<Volume> = vec![];
+        for volume in grpc_volume_type.volumes {
+            volumes.push(Volume::try_from(volume.clone())?)
```

Comment:
don't need to clone here if you consume grpc_volume_type? Also I'd consider renaming grpc_volume_type to grpc_volumes or something like that?